Modules/Registration/Metricsv4/test/itkEuclideanDistancePointSetMetricTest3.cxx

```diff
@@ -0,0 +1,252 @@
+  typename PointSetMetricType::DerivativeType derivative, derivative2;
+  metric->GetDerivative(derivative);
+  metric->GetValueAndDerivative(value2, derivative2);
+
+  std::cout << "value: " << value << std::endl;
+
+  // Check for the same results from different methods
+  if (itk::Math::NotExactlyEquals(value, value2))
+  {
+    std::cerr << "value does not match between calls to different methods: "
+              << "value: " << value << " value2: " << value2 << std::endl;
+  }
+  if (derivative != derivative2)
+  {
+    std::cerr << "derivative does not match between calls to different methods: "
+              << "derivative: " << derivative << " derivative2: " << derivative2 << std::endl;
+  }
+
+  // Check if the points outside threshold are skipped in metric calculation
+  double distanceSum = 0.0;
+  for (unsigned int n = 0; n < numberOfPoints; ++n)
+  {
+    if (distanceThreshold <= 0 || distanceArray[n] < distanceThreshold)
+    {
+      distanceSum = distanceSum + distanceArray[n];
+    }
+  }
+
+  double valueTest = distanceSum / numberOfPoints;
+  if (itk::Math::NotExactlyEquals(valueTest, value2))
+  {
+    std::cerr << "Value calculation is wrong when used threshold : " << distanceThreshold << "valueTest: " << valueTest
+              << " value2: " << value2 << std::endl;
+    return EXIT_FAILURE;
+  }
+
+
+  // Check if the point outside threshold is skipped in derivative calculation
+  typename PointSetMetricType::DerivativeType derivativeTest;
+  derivativeTest.SetSize(Dimension);
+  derivativeTest.Fill(0);
+  for (unsigned int n = 0; n < numberOfPoints; ++n)
+  {
+    auto tempFixedPoint = fixedPoints->GetPoint(n);
+    auto tempMovingPoint = movingPoints->GetPoint(n);
+    auto tempDerivative = tempMovingPoint - tempFixedPoint;
+
+    if (distanceThreshold <= 0 || distanceArray[n] < distanceThreshold)
+    {
+      derivativeTest[0] = derivativeTest[0] + tempDerivative[0];
+      derivativeTest[1] = derivativeTest[1] + tempDerivative[1];
+      if (Dimension == 3)
+      {
+        derivativeTest[2] = derivativeTest[2] + tempDerivative[2];
+      }
+    }
+  }
+
+  auto derivativeTestMean = derivativeTest / numberOfPoints;
+  std::cout << "Derivative is [ " << derivativeTestMean << " ]" << std::endl;
+
+  for (unsigned int i = 0; i << Dimension; ++i)
+  {
+    if (itk::Math::NotExactlyEquals(derivativeTestMean[i], derivative2[i]))
+    {
+      std::cerr << "Derivative calculation is wrong when used threshold : " << distanceThreshold
+                << "derivativeTestMean: " << derivativeTestMean << " derivative2: " << derivative2 << std::endl;
+      return EXIT_FAILURE;
+    }
+  }
+
+  return EXIT_SUCCESS;
+}
+
+int
+itkEuclideanDistancePointSetMetricTest3(int, char *[])
+{
+  int result = EXIT_SUCCESS;
+
+  // Test for positive distance threshold
+  if (itkEuclideanDistancePointSetMetricTest3Run<2>(0.5) == EXIT_FAILURE)
+  {
+    std::cerr << "Failed for Dimension 2." << std::endl;
```

Comment:
The message could be more specific; now they do not allow knowing which case caused an error
```
constexpr unsigned int Dimension2D = 2:
double distanceThreshold = 0.5;
(...)
std::cerr << "Failed for Dimension " << Dimension2D << " and distance threshold " << distanceThreshold << std::endl;
```